nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py

```diff
@@ -0,0 +1,350 @@
+
+        position_embeddings = self.frozen_model.enc_dec_model.encoder_embedding.position_embeddings(position_ids)
+        encoder_input = input_embeds + position_embeddings
+
+        predicted_token_ids, log_probs = self.frozen_model.decode(
+            tokens_enc=enc_input,
+            enc_mask=enc_mask,
+            num_tokens_to_generate=self.decoder_seq_length,
+            encoder_input=encoder_input,
+        )
+
+        return {
+            'enc_input': enc_input,
+            'predicted_token_ids': predicted_token_ids,
+            'log_probs': log_probs,
+            'labels': labels,
+        }
+
+    def on_predict_epoch_end(self, outputs: List[Any]) -> None:
+
+        all_preds = []
+        all_labels = []
+        all_inputs = []
+        for item in outputs[0]:
+            preds = item['predicted_token_ids'].cpu().numpy().tolist()
+            enc_inputs = item['enc_input'].cpu().numpy().tolist()
+            if item['labels'] is not None:
+                labels = item['labels'].cpu().numpy().tolist()
+            else:
+                labels = [None] * len(preds)
+
+            for i, (enc_input, pred, label) in enumerate(zip(enc_inputs, preds, labels)):
+                if self.tokenizer.eos_id in pred:
+                    idx = pred.index(self.tokenizer.eos_id)
+                    pred = pred[:idx]
+
+                pred = [
+                    id
+                    for id in pred
+                    if id not in self.tokenizer.tokenizer.additional_special_tokens_ids
+                    and id not in self.tokenizer.text_to_ids(T5Sentinel.FIRST.value)
+                ]  # delete the sentinel token at the beginning of prediction
+                pred = self.tokenizer.ids_to_text(pred)
+                all_preds.append(pred)
+
+                enc_input = [
+                    id for id in enc_input if id not in self.tokenizer.text_to_ids(T5Sentinel.FIRST.value)
+                ]  # delete the sentinel token added to the end of input
+                input = self.tokenizer.ids_to_text(enc_input)
+                all_inputs.append(input)
+
+                # If labels are given, collect them to calculate the accuracy
+                if label is not None:
+                    label = [
+                        id
+                        for id in label
+                        if id not in self.tokenizer.tokenizer.additional_special_tokens_ids
+                        and id not in self.tokenizer.text_to_ids(T5Sentinel.FIRST.value)
+                    ]  # delete the sentinel token at the beginning of label
+                    label = self.tokenizer.ids_to_text(label)
+                    all_labels.append(label)
+
+        correct, acc = 0, None
+        if all_labels:
+            for pred, label in zip(all_preds, all_labels):
+                if pred == label:
+                    correct += 1
+            acc = correct / len(all_preds)
```

Comment:
If your data parallel size is > 1 at predict time, this will only compute accuracy per data parallel rank. Since you use the DistributedSampler, this will only be the accuracy on a subset of your dataset. You can either use `TorchMetrics` like I do in the t5 finetuning class or do an allgather object from all data parallel ranks like we do in NMT (see https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py#L337)